### ISSUE TITLE
WebSockets/Channels support

### ISSUE THREAD

**Issue by flux627** [Miserlou/Zappa#649]
This isn't as much of a feature request as it is an attempt to start of a conversation about the possibility of adding WebSocket support (through Channels or otherwise). This would probably have to be done through AWS IoT ([MQTT over WebSockets](https://aws.amazon.com/about-aws/whats-new/2016/01/aws-iot-now-supports-websockets-custom-keepalive-intervals-and-enhanced-console/)). 

Here is a proof-of-concept using this with AWS Lambda:

http://stesie.github.io/2016/04/aws-iot-pubsub

I also found this, but I'm not sure if it helps us any:

https://github.com/edigiacomo/asgi-mqtt

**Comment by pjz** [comment]
I don't see a way to get plain Websocket support, only MQTT-over-websockets.  And you don't really need Lambda for that.

**Comment by flux627** [comment]
I think having a websocket solution that can work within the context of this framework would be valuable, however I do see your point in that it doesn't have to be directly coupled with Zappa. Maybe this would be better suited as a separate library?

**Comment by jgissend10** [comment]
Bumping this as websockets are now a part of api-gateway!
https://aws.amazon.com/blogs/compute/announcing-websocket-apis-in-amazon-api-gateway/

**Comment by makarovstas** [comment]
> Bumping this as websockets are now a part of api-gateway!
> https://aws.amazon.com/blogs/compute/announcing-websocket-apis-in-amazon-api-gateway/

Hi guys!  Any ideas, how implement this to zappa?

**Comment by mojimi** [comment]
For this to work, there would need to be a Django/Flask Websocket extension that worked like AWS API Gateway, it has a specific way of dealing the connection, working as a middleware

**Comment by mojimi** [comment]
In terms of syntax/abstraction, I think a great way to implement this would be by extending zappa-async to handle the web socket logic.

**Comment by erm** [comment]
I am not certain if this would be the solution here, but I've created an adapter library for ASGI on AWS Lambda & API Gateway called [here](https://github.com/erm/mangum) with an open issue about WebSocket support [here](https://github.com/erm/mangum/issues/1). I haven't had much time to dig into the WebSocket support specifically, but I'm very interested in finding the solution.

If there is an interest in supporting ASGI in Zappa I would be happy to assist where I can, in which case I think to approach from HTTP support first and then tackling WebSockets after would be the way to go.

**Comment by jonbesga** [comment]
Is there any update on this topic?

**Comment by delijati** [comment]
I added it in my Zappa branch https://github.com/delijati/Zappa. Here i played a bit around with it https://github.com/delijati/benchmark :)

**Comment by lukebranch** [comment]
> I added it in my Zappa branch https://github.com/delijati/Zappa. Here i played a bit around with it https://github.com/delijati/benchmark :)

Nice work! Have you played around with any ASGI frameworks (eg. https://bocadilloproject.github.io/) with your fork of Zappa?

**Comment by pjz** [comment]
or quart?  Now that this is possible, I'm definitely :+1: on it!

**Comment by delijati** [comment]
the benchmark uses `quart` ;) i really would like to know why the WSGI seams faster

**Comment by arvinsim** [comment]
Any update on this? Seems to be bummer for people working on ASGI frameworks.

**Comment by monkut** [comment]
Yes, I'm like what I'm seeing in [fastapi](https://fastapi.tiangolo.com/), what's the status here, any clear direction  or work that needs to be done?

**Comment by jdhayes01** [comment]
Any update on this? Im trying to implement websockets with zappa and AWS Lambda

**Comment by jneves** [comment]
Has anyone tried out the api gateway support for websockets? What doesn't work? https://aws.amazon.com/blogs/compute/announcing-websocket-apis-in-amazon-api-gateway/

**Comment by acidjunk** [comment]
Is there already a way to get this working somehow? Also interested in proof of concepts, suggestions, workarounds, e.g. any info at all.

**Comment by jordaneremieff** [comment]
@acidjunk 

I was working on an ASGI protocol implementation to support WebSockets on a [branch](https://github.com/jordaneremieff/mangum/blob/broadcast/mangum/protocols/websockets.py) in a different project, though it may be applicable here if ASGI support is introduced. I was able to create a successful deployment in an [example](https://github.com/jordaneremieff/mangum-chatroom/blob/main/src/asgi.py) application (based on a previous commit in that branch). The implementation follows the backing-service approach demonstrated in [broadcaster](https://github.com/encode/broadcaster) to persist the connection identifiers.

I included detailed docstrings/comments in the protocol class to explain how the flow works. There is also more information on the old doc [page](https://github.com/jordaneremieff/mangum/blob/broadcast/docs/websockets.md) and in this [issue](https://github.com/django/asgiref/issues/169). 

Essentially the way it works:

- A $connect event is received, and the adapter stores the connection id in the backing service and returns successfully to complete the handshake.

- Any $message events fetch the connection id from the backing-service and runs a new ASGI protocol+application instance until the connection becomes stale or the application decides to close it.

- A $disconnect event should remove the connection id from the backing-service.

I'm tracking future progress on any of my future efforts [here](https://github.com/jordaneremieff/mangum/issues/132), but for the moment I am bored of this problem and haven't been motivated to push forward. If you decide to take a shot at this and have any specific questions, feel free to comment on that issue.

Hope any of this helps! :)

**Comment by deceze** [comment]
API Gateway Websocket integration should be pretty straight forward:

- API Gateway handles the persistency, so the Lambda handler/Zappa can remain stateless
- API Gateway triggers events, which can invoke Lambda function responses
- pushing messages to a socket can be done via boto (e.g. `boto3.client('apigatewaymanagementapi').post_to_connection(ConnectionId=connection_id, Data=...)`)

In other words, incoming events can be handled like any other HTTP request or scheduled event, and outgoing data is a simple HTTP request as well.

Currently this is best implemented by setting up an HTTP Proxy integration in the Websocket API Gateway, triggering the Lambda handler via HTTP. What Zappa could do here would be to augment the `LambdaHandler` class to accept API Gateway Websocket events directly as well; currently those simply don't fit any of the existing patterns and just fall through, doing nothing. Then a Lambda function integration could be done, which is probably a bit more performant. Integrating some templates for making that API Gateway setup a simple configuration option would be nice too.

**Comment by ganeshprasadrao** [comment]
Guys, Any update on this?

**Comment by pfcodes** [comment]
Bump

**Comment by souravjamwal77** [comment]
Hi @mister-rao and @pfcodes,
Can you please ask for this feature on currently active repo here [https://github.com/zappa/Zappa/issues/949](link) ?
